### DISCUSSION TITLE
Mogrify: Applying both "background" and "fill" to the same image

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#6907] [Help]
### ImageMagick version

7.1.1 Latest

### Operating system, version and so on

Windows 11

### Description

I'm trying to find a way to change a Font Awesome SVG in to a PNG with both a background and a fill. My main command is:

mogrify -format png -path "Converted" -gravity center -size 40x40 -extent 40x40 [BACKGROUND, FILL, AND COLORIZE HERE] *.*

Now here's the different background, fill and colorize arguments I've used and their result (1BFF00 is green, FF0000 is red):

-background none -fill #FF0000 | This was a test - the fill colour didn't apply
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/f1b38302-57b5-4f11-beac-9acb10a54466)

-background none -fill #FF0000 -colorize 100 | This was a test - the fill colour worked and the icon is red
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/a000a51f-665b-4e10-8dd0-549d351911d2)

-background 1BFF00 -colorize 50 -fill #FF0000 -colorize 50 | This was a test - each colour can be seen at its 50% opacity
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/f22a1ca5-8965-4a45-9ea2-f29d9e1b5ddc)

-background 1BFF00 -colorize 100 -fill #FF0000 | Green image
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/cf57e537-6cd4-4b35-85cf-804b28649756)

-background 1BFF00 -colorize 100 -fill #FF0000 -colorize 100 | Red image
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/54141f38-d090-4ec2-8351-6ceaee38d5be)

-background 1BFF00 -fill #FF0000 | The background worked but the fill didn't
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/76936cdb-747b-4f4c-b5a0-1e84d49eb34e)

-background 1BFF00 -fill #FF0000 -colorize 100 | Red image
![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/5f41af35-79ef-4d61-9419-2a3044982233)

I've tried the documentation, Google, and searching here, and I am at a loss of how to comine background and fill.

Thanks for any help.

**Comment by GeeMack** [comment]
The "-colorize" operation uses the "-fill" color setting, but "-fill" doesn't affect the background operations. The "-background" sets a color to use as a background when needed for certain append, extent, splice, etc. operations. To get your font red and your background green, try something like thjis..

    ... -fill #FF0000 -colorize 100 -background #1BFF00 -flatten ...

**Reply by ghost** [reply]
Thanks for your reply. Indeed I saw on the docs that colorize applies to fill, but I was confused by the different behaviour I was getting in testing. I've given your command a go, in that order and the reverse/original order (background then fill) and I'm still getting the same red filled square

mogrify -format png -path "Converted" -gravity center -size 40x40 -extent 40x40 -fill #FF0000 -colorize 100 -background #1BFF00 -flatten *.*

![amazon](https://github.com/ImageMagick/ImageMagick/assets/89745850/c201862c-f307-426c-9c1d-4230ef3103f5)

**Reply by GeeMack** [reply]
Your "-extent" operation changes the background from transparent to white, ImageMagick's default background color. Then the red "-colorize" covers the whole image. Try first `-fill #FF0000 -colorize 100` to make just the font image red. Then set the background color. Then when you do the "-extent" operation it should also place it on the background color.

    ... -fill #FF0000 -colorize 100 -background #1BFF00 -extent 40x40 ...

**Reply by ghost** [reply]
Interesting, I didn't know the extent would do that, although sadly it still isn't working... now I'm getting green parts on the part of the image that was extented

mogrify -format png -path "Converted" -gravity center -size 40x40 -fill #FF0000 -colorize 100 -background #1BFF00 -extent 40x40 *.*

Examples of first 10 icons:

![adversal](https://github.com/ImageMagick/ImageMagick/assets/89745850/b401ad2c-c05a-4e77-b114-5bbcd6e4f888)
![affiliatetheme](https://github.com/ImageMagick/ImageMagick/assets/89745850/e01a2a1b-3190-467b-8e67-7c1ab710c6c7)
![airbnb](https://github.com/ImageMagick/ImageMagick/assets/89745850/2c2d9c0a-420a-4954-8a7d-10cc59ec9186)
![algolia](https://github.com/ImageMagick/ImageMagick/assets/89745850/9f293ef1-b21f-4441-af83-df4ed06e7fe9)
![alipay](https://github.com/ImageMagick/ImageMagick/assets/89745850/1470bd23-f008-4cc9-99af-c4ab109db159)
![42-group](https://github.com/ImageMagick/ImageMagick/assets/89745850/a716af96-5e31-44aa-b873-37852126304b)
![500px](https://github.com/ImageMagick/ImageMagick/assets/89745850/e041da0a-7d2b-435b-a8b2-9fb56141d7b7)
![accessible-icon](https://github.com/ImageMagick/ImageMagick/assets/89745850/e47cfc1b-5692-4a29-b864-16b2cef70238)
![accusoft](https://github.com/ImageMagick/ImageMagick/assets/89745850/7ced289f-22d9-4517-81df-b7a8ac1516c3)
![adn](https://github.com/ImageMagick/ImageMagick/assets/89745850/9682eeea-bb2f-4873-b431-7e22c4a3d5df)

**Comment by snibgo** [comment]
> ... now I'm getting green parts on the part of the image that was extented

That is because you have specified a green colour for the extent:
```
-background #1BFF00 -extent 40x40
```
`-background` and `-fill` are settings. They do not change the image. They do affect how some operations change the image.

In your case, `-background` affects `-extent`, and `-fill` affects `-colorize`.
